sfge/src/main/java/com/salesforce/rules/ops/boundary/SharingPolicyBoundaryDetector.java

```diff
@@ -0,0 +1,107 @@
+            throw new ProgrammingException(
+                    "SharingPolicyBoundaryDetector tried to find "
+                            + "the current sharing boundary but the boundary stack is empty.");
+        }
+        return super.boundaries.peek();
+    }
+
+    /**
+     * Find the currently effective sharing boundary, possibly deeper in the stack if the first
+     * boundary has "inherited sharing" or no explicitly defined sharing policy. Note: the stack
+     * must not be empty.
+     *
+     * @return a {@link SharingPolicyBoundary} where {@link
+     *     SharingPolicyBoundary#getApplicableAncestorPolicyVertex()} returns a {@link
+     *     UserClassVertex} with an explicitly defined sharing policy. If there are none, returns
+     *     the {@link SharingPolicyBoundary} at the top of the stack.
+     */
+    public SharingPolicyBoundary getEffectiveSharingBoundary() {
+        if (super.boundaries.isEmpty()) {
+            throw new ProgrammingException(
+                    "SharingPolicyBoundaryDetector tried to find "
+                            + "the current sharing boundary but the boundary stack is empty.");
+        }
+        final Stack<SharingPolicyBoundary> copyStack = new Stack<>();
+        copyStack.addAll(super.boundaries);
+        return this.__getCurrentSharingPolicy(copyStack);
+    }
+
+    /** Get the currently effective sharing policy boundary, recursively */
+    private SharingPolicyBoundary __getCurrentSharingPolicy(
+            Stack<? extends SharingPolicyBoundary> boundaryStack) {
+
+        /*
+         * Note: the currently effective sharing policy can come from a few places
+         *
+         * Option 1: the current class has an explicitly defined sharing policy that is
+         * "with sharing" or "without sharing". Use that.
+         *
+         * Option 2: the current class has no explicitly defined sharing policy, but it
+         * has an ancestor with a "with sharing" or "without sharing". Use the ancestor's
+         * policy.
+         *
+         * Option 3: the current class has no explicitly defined sharing policy, but it
+         * has an ancestor with an "inherited sharing". Look deeper in the stack to inherit
+         * the most recently explicitly defined sharing policy.
+         *
+         * Option 4: the current class has no explicitly defined sharing policy, and neither
+         * do its ancestors. Look deeper in the boundary stack to implicitly inherit the most
+         * recently explicitly defined sharing policy.
+         *
+         */
+
+        final Optional<? extends SharingPolicyBoundary> sharingPolicyBoundaryOpt =
+                CollectionUtil.peek(boundaryStack);
+
+        if (!sharingPolicyBoundaryOpt.isPresent()) {
+            throw new ProgrammingException(
+                    "SharingPolicyBoundaryDetector cannot find current sharing policy on empty policy stack.");
+        }
+        boundaryStack.pop();
```

Comment:
Is there a reason you can't use `.empty()` and `pop()` instead of `peek` and an `Optional`?